nixos/lib/test-driver/test_driver/qmp.py

```diff
@@ -0,0 +1,65 @@
+import socket
+from queue import Queue
+from typing import Dict, Any, Iterator
+import json
+from pathlib import Path
+
+
+class QMPSession:
+    def __init__(self, sock: socket.socket) -> None:
+        self.sock = sock
+        self.pending_events: Queue[Dict[str, Any]] = Queue()
+        self.reader = sock.makefile("r")
+        self.writer = sock.makefile("w")
+        hello = self._result()
+        assert "QMP" in hello, f"Unexpected result: {hello}"
+        self.send("qmp_capabilities")
+
+    @classmethod
+    def from_path(cls, path: Path) -> "QMPSession":
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.connect(str(path))
+        return cls(sock)
+
+    def __del__(self) -> None:
+        self.sock.close()
+
+    def _readmsg(self) -> Dict[str, Any]:
+        line = self.reader.readline()
+        return json.loads(line)
+
+    def _raise_unexpected_msg(self, msg: Dict[str, Any]) -> None:
+        m = json.dumps(msg, sort_keys=True, indent=4)
+        raise RuntimeError(f"Got unexpected qmp response: {m}")
+
+    def _result(self) -> Dict[str, Any]:
+        while True:
+            # QMP is in the handshake
+            res = self._readmsg()
+            if "return" in res or "QMP" in res:
+                return res
+            elif "event" in res:
+                self.pending_events.put(res)
+                continue
+            else:
+                self._raise_unexpected_msg(res)
+
+    def events(self) -> Iterator[Dict[str, Any]]:
+        while not self.pending_events.empty():
+            yield self.pending_events.get()
+
+        res = self._readmsg()
+
+        if "event" not in res:
```

Comment:
it seems unfortunate that we have two different codepaths that use the _readmsg function.

Can this please be refactored in a way that there is a function that for example shakes the tree for new outputs, that is the only one which implements knowledge about what kind of keys can be expected in the QMP message outputs, and then have that be called from all the others?

Reply:
I tried something, let me know how you like it.